Source/CesiumRuntime/Private/CesiumCreditSystem.cpp

```diff
@@ -163,45 +175,125 @@ void ACesiumCreditSystem::Tick(float DeltaTime) {
-    if (hasPopupCredits) {
-      // create phony url to detect when to open popup menu
-      onScreenCreditString +=
-          "<span> </span><a href=\"cesium-unreal://data-attribution-popup\">Data attribution</a>";
-    }
-    onScreenCreditString += "</body>')";
-    OnScreenCredits = UTF8_TO_TCHAR(onScreenCreditString.c_str());
-    _lastCreditsCount = creditsToShowThisFrame.size();
+    OnScreenCredits += "<credits url=\"popup\" text=\" Data attribution\"/>";
+    _creditsWidget->SetCredits(Credits, OnScreenCredits);
+  }
+  _pCreditSystem->startNextFrame();
+}
 
-    DisplayCredits = hasPopupCredits || firstScreenCredit;
+namespace {
+void convertHtmlToRtf(
+    std::string& output,
+    std::string& parentUrl,
+    TidyDoc tdoc,
+    TidyNode tnod,
+    UScreenCreditsWidget* CreditsWidget) {
+  TidyNode child;
+  for (child = tidyGetChild(tnod); child; child = tidyGetNext(child)) {
+    if (tidyNodeIsText(child)) {
+      TidyBuffer buf;
+      tidyBufInit(&buf);
+      tidyNodeGetText(tdoc, child, &buf);
+      if (buf.bp) {
+        std::string text = reinterpret_cast<const char*>(buf.bp);
+        // could not find correct option in tidy html to not add new lines
+        if (text.size() != 0 && text[text.size() - 1] == '\n') {
+          text.pop_back();
+        }
+        if (!parentUrl.empty()) {
+          output +=
+              "<credits url=\"" + parentUrl + "\"" + " text=\"" + text + "\"/>";
+        } else {
+          output += text;
+        }
+      }
+      tidyBufFree(&buf);
+    } else if (tidyNodeGetId(child) == TidyTagId::TidyTag_IMG) {
+      auto srcAttr = tidyAttrGetById(child, TidyAttrId::TidyAttr_SRC);
+      if (srcAttr) {
+        auto srcValue = tidyAttrValue(srcAttr);
+        if (srcValue) {
+          output += "<credits id=\"" +
+                    CreditsWidget->LoadImage(
+                        std::string(reinterpret_cast<const char*>(srcValue))) +
+                    "\"";
+          if (!parentUrl.empty()) {
+            output += " url=\"" + parentUrl + "\"";
+          }
+          output += "/>";
+        }
+      }
+    }
+    auto hrefAttr = tidyAttrGetById(child, TidyAttrId::TidyAttr_HREF);
+    if (hrefAttr) {
+      auto hrefValue = tidyAttrValue(hrefAttr);
+      parentUrl = std::string(reinterpret_cast<const char*>(hrefValue));
+    }
+    convertHtmlToRtf(output, parentUrl, tdoc, child, CreditsWidget);
   }
+}
+} // namespace
 
-  _pCreditSystem->startNextFrame();
+FString ACesiumCreditSystem::ConvertHtmlToRtf(std::string html) {
+  TidyDoc tdoc;
+  TidyBuffer docbuf = {0};
+  TidyBuffer tidy_errbuf = {0};
+  int err;
+
+  tdoc = tidyCreate();
+  tidyOptSetBool(tdoc, TidyForceOutput, yes);
+  tidyOptSetInt(tdoc, TidyWrapLen, 0);
+  tidyOptSetInt(tdoc, TidyNewline, TidyLF);
+
+  tidySetErrorBuffer(tdoc, &tidy_errbuf);
+  tidyBufInit(&docbuf);
+
+  // avoid buffer overflow by wrapping with html tags
+  html = "<span>" + html + "</span>";
```

Comment:
It's strange that this is necessary. I tried commenting it out, and it looks like it crashes while trying to build an error message that says `plain text isn't allowed in <head> elements`. Perhaps the crash is a bug in the development version of tidy-html5 we're using? I'm not really sure how it's meant to work when we parse a HTML document that isn't really a document, but is more like an HTML fragment. If that's the problem - credits are being interpreted as part of `<head>` and that's sometimes invalid - maybe the right approach is to wrap it up to be a proper document:

```
<!DOCTYPE html>
<html>
<body>
{{credit text goes here}}
</body>
</html>
```

Probably not much different from adding the `<span>`, but perhaps more justifiable?